app/javascript/dashboard/components/widgets/conversation/helpers/SLAHelper.js

```diff
@@ -0,0 +1,135 @@
+    formatTime(hours, 'h', minutes, 'm') ||
+    `${minutes > 0 ? minutes : 1}m`
+  );
+};
+
+// Format the SLA time in the format of years, months, days, hours and minutes, eg. 1y 2mo 3d 4h 5m
+export const formatSLATime = time => {
+  if (time < 0) return '';
+
+  const minutes = convertToMinutes(time);
+  const [hours, remainingMinutes] = convertToHoursAndMinutes(minutes);
+  const [days, remainingHours] = convertToDaysAndHours(hours);
+  const [months, remainingDays] = convertToMonthsAndDays(days);
+  const [years, remainingMonths] = convertToYearsAndMonths(months);
+
+  return formatDuration(
+    years,
+    remainingMonths,
+    remainingDays,
+    remainingHours,
+    remainingMinutes
+  );
+};
+
+const createSLAObject = (
+  type,
+  {
+    first_response_time_threshold,
+    next_response_time_threshold,
+    resolution_time_threshold,
+  } = {},
+  {
+    first_reply_created_at: firstReplyCreatedAt,
+    waiting_since: waitingSince,
+    created_at: createdAt,
+    status,
+  } = {}
+) => {
+  // Mapping of breach types to their logic
+  const SLATypes = {
+    FRT: {
+      threshold: calculateThreshold(createdAt, first_response_time_threshold),
+      //   Check FRT only if threshold is not null and first reply hasn't been made
+      condition:
+        first_response_time_threshold !== null &&
+        (!firstReplyCreatedAt || firstReplyCreatedAt === 0),
+    },
+    NRT: {
+      threshold: calculateThreshold(waitingSince, next_response_time_threshold),
+      // Check NRT only if threshold is not null, first reply has been made and we are waiting since
+      condition:
+        next_response_time_threshold !== null &&
+        !!firstReplyCreatedAt &&
+        !!waitingSince,
+    },
+    RT: {
+      threshold: calculateThreshold(createdAt, resolution_time_threshold),
+      // Check RT only if the conversation is open and threshold is not null
+      condition: status === 'open' && resolution_time_threshold !== null,
+    },
+  };
+
+  const SLAStatus = SLATypes[type];
+  return SLAStatus ? { ...SLAStatus, type } : null;
+};
+
+const evaluateSLAConditions = (sla, chat) => {
+  // Filter out the SLA based on conditions and update the object with the breach status(icon, isSlaMissed)
+  const SLATypes = ['FRT', 'NRT', 'RT'];
+  return SLATypes.map(type => createSLAObject(type, sla, chat))
+    .filter(SLAStatus => SLAStatus && SLAStatus.condition)
+    .map(SLAStatus => ({
+      ...SLAStatus,
+      icon: SLAStatus.threshold <= 0 ? 'flame' : 'alarm',
+      isSlaMissed: SLAStatus.threshold <= 0,
+    }));
+};
+
+export const evaluateSLAStatus = (sla, chat) => {
+  if (!sla || !chat)
+    return { type: '', threshold: '', icon: '', isSlaMissed: false };
+
+  // Filter out the SLA and create the object for each breach
+  const SLAStatuses = evaluateSLAConditions(sla, chat);
+
+  // Return the most urgent SLA which is latest to breach or has breached
+  const mostUrgent = findMostUrgentSLAStatus(SLAStatuses);
+  return mostUrgent
```

Comment:
@sojan-official is incorporating SLA events into the conversation. This will let us directly extract breached events from the conversation instead of calculating them separately. This method will also help us identify which SLA events have been breached. We only need to determine the next breaching time. By doing that we can simplify this logic as well.

Reply:
Closing this comment, as discussed, we are not using the sla_event to find out the logic here.